### PR TITLE
New dhcp state machine

### LINKED ISSUE #94: DHCP Finite State Machine mishandles renew and rebind.
Issues with current DHCP FSM:
1) Only makes one attempt to obtain a lease and then waits for timeout. UDP is lossy and DHCP standard calls for an exponential fall back retry strategy.
2) When calling the DHCP maintenance routine, the first time it is called after timer T1 has expired, it does not attempt to renew the lease, only sets the FSM state to START. Lease renewal will then be attempted on the next call.
3) If renewal fails (because the original DHCP server is not responding) and timer T2 has expired, FSM does not attempt to rebind by broadcasting the DHCP request, it continues to attempt to renew.
4) Sends messages with broadcast instead of unicast (the DHCP standard to reduce spurious network traffic).

### PR DESCRIPTION
**Pull request by BiffoBear**
closes #94 New DHCP FSM implements exponential fallback and correctly renews or rebinds. The messages are sent  in unicast mode. Also added a debug message helper function with a hexdump to improve code readability. DHCP messages are now sent and received without using the socket module. This should save some memory.

### CODE REVIEW

**Comment by brentru** [comment]
@BiffoBear Thank you for the PR! I'll review it shortly.

@FoamyGuy Do you have time this upcoming week to test this on hardware?

File: adafruit_wiznet5k/__init__.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    if TYPE_CHECKING:
+        from adafruit_wiznet5k.adafruit_wiznet5k import WIZNET5K
+except ImportError:
+    pass
+
+import gc
+
+
+def debug_msg(
```
Comment (brentru) [inline adafruit_wiznet5k/__init__.py:17]: I'd prefer if __init__.py was only used for initialization and configuration, not to declare methods.

Could you instead pull the helper functions into `_wiznet5k.py` or another file?

File: adafruit_wiznet5k/adafruit_wiznet5k.py [hunk 2]
```diff
@@ -61,6 +62,10 @@
 _REG_SIPR = const(0x000F)  # Source IP Address
 _REG_PHYCFGR = const(0x002E)  # W5500 PHY Configuration
 _REG_PHYCFGR_W5100S = const(0x003C)  # W5100S PHY Configuration
+_REG_RCR_5100s = const(0x0019)  # Retry Count
```
Comment (brentru) [inline adafruit_wiznet5k/adafruit_wiznet5k.py:65]: Q: You removed `dhcp_timeout` from the class initialization method in favor of hardcoding these values. Will users _always_ want to use this retry time/retry count, or is there a future need for it to be set-able by a user?

File: adafruit_wiznet5k/adafruit_wiznet5k_dns.py [hunk 3]
```diff
@@ -275,7 +273,7 @@ def gethostbyname(self, hostname: bytes) -> Union[int, bytes]:
                     return -1
                 time.sleep(0.05)
             # recv packet into buf
-            buffer = self._sock.recv(512)  # > UDP payload length
+            buffer = self._sock.recv()
```
Comment (brentru) [inline adafruit_wiznet5k/adafruit_wiznet5k_dns.py:276]: Why are we removing the 512-byte UDP payload length being set here? Is there no longer a limit for how much we want to recv?

**Review by brentru — CHANGES_REQUESTED** [review]
Really great work here, thank you for the pull request. Few questions!

File: adafruit_wiznet5k/__init__.py (hunk 1 above)
Comment (BiffoBear) [reply adafruit_wiznet5k/__init__.py:17]: I'll move it to something like `wiznet5k_debug.py` then. It'll be imported by each module that has debugging, eventually.

File: adafruit_wiznet5k/adafruit_wiznet5k.py (hunk 2 above)
Comment (BiffoBear) [reply adafruit_wiznet5k/adafruit_wiznet5k.py:65]: The `dhcp_timeout` set by the user was used to decide how long to wait for a response to a DHCP request from the server. This is obsolete as the time out now follows the DHCP standard of  2 +/- 1 seconds, 4 +/- 1 seconds, etc up to 32 +/- 1 seconds. The new retry time and count is used to implement that standard. I don't think this is something the user needs to adjust, but I'm happy to discuss possible use cases.

Comment (brentru) [reply adafruit_wiznet5k/adafruit_wiznet5k.py:65]: > This is obsolete as the time out now follows the DHCP standard

Sweet, thank you for the clarification.

File: adafruit_wiznet5k/adafruit_wiznet5k_dns.py (hunk 3 above)
Comment (BiffoBear) [reply adafruit_wiznet5k/adafruit_wiznet5k_dns.py:276]: I have no idea TBH. I'll revert it to `recv(512)` to match the current `main` branch. Great catch. Thank you!

File: adafruit_wiznet5k/__init__.py (hunk 1 above)
Comment (BiffoBear) [reply adafruit_wiznet5k/__init__.py]: OK, I've moved it from `init.py` to `adafruit_wiznet5k.adafruit_wiznet5k_debug.py`

File: adafruit_wiznet5k/adafruit_wiznet5k_dns.py (hunk 3 above)
Comment (brentru) [reply adafruit_wiznet5k/adafruit_wiznet5k_dns.py]: thx!

**Review by brentru — APPROVED** [review]

**Comment by FoamyGuy** [comment]
> @FoamyGuy Do you have time this upcoming week to test this on hardware?

Yep, I'll try this out today.

**Comment by FoamyGuy** [comment]
The version from this PR doesn't seem to work as well as the currently released version in my network environment. 
Testing was performed with Ethernet Featherwing and Feather ESP32-S2 TFT w/ version 8.0.0.

I'm seeing two things that seem to be problems with the new version:

1) There is a byte / str conversion error reported when I try to run the simpletest in this repo:
```
Wiznet5k WebClient Test
Chip Version: w5500
MAC Address: ['0xde', '0xad', '0xbe', '0xef', '0xfe', '0xed']
My IP address is: 192.168.1.208
Traceback (most recent call last):
  File "code.py", line 33, in <module>
  File "/lib/adafruit_wiznet5k/adafruit_wiznet5k.py", line 264, in get_host_by_name
  File "/lib/adafruit_wiznet5k/adafruit_wiznet5k_dns.py", line 255, in gethostbyname
  File "/lib/adafruit_wiznet5k/adafruit_wiznet5k_socket.py", line 354, in connect
TypeError: can't convert 'str' object to bytes implicitly
```
I tried a few different attempts and I seem to get this error raised consistently (when DHCP succeeds after a cold boot). 

Was there a change to the types expected to be returned by `get_host_by_name()` or anything in this chain that would result in the code in the example needing to be changed? This is the line that raises the exception:
```
print(
    "IP lookup adafruit.com: %s" % eth.pretty_ip(eth.get_host_by_name("adafruit.com"))
)
```

2) With this version I am unable to get multiple successful DHCP negotiations in a row with Ctrl C / Ctrl D in between to reset. As far as I can tell this is unrelated to issue 1 above. Immediately after a cold boot up the DHCP can succeed. But pressing Ctrl C/D after that to reset consistently results in a DHCP failure no matter how many times I try until I do a full unplug / re-plug cold boot cycle.  On the currently released version I can get multiple successful DHCP results in a row by using ctrl C/D in between.

This is a debug=True log output from a successful DHCP that occurred immediately after a cold boot:  [successful_dhcp_after_cold_boot.log](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/10724340/successful_dhcp_after_cold_boot.log)

This is a debug=True log output from an unsuccessful DHCP that occurs after ctrl C/D reset. I get this same outcome over and over with successive attempts at ctrl c/d resets. It only goes back to success after a full unplug / replug boot cycle: 
[unsuccessful_dhcp_after_ctrl_cd.log](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/10724354/unsuccessful_dhcp_after_ctrl_cd.log)

**Comment by BiffoBear** [comment]
@FoamyGuy Hi, Thanks for testing. I'll try to replicate what you're seeing on your network. I like it when you test Wiznet5k because you are clearly in a very different environment to me 😀

If I cannot replicate these problems, may I then ask you for a WireShark file so that I can see the raw conversation?

Thanks again

**Comment by FoamyGuy** [comment]
My environment definitely seems to be good at breaking things :sweat_smile:. 

I am willing to try to capture whatever logs I can to help troubleshoot. I haven't used wireshark for a long time and not at all on my current OS. I'll have to figure out how to get it set up and capturing things.

**Comment by BiffoBear** [comment]
> My environment definitely seems to be good at breaking things 😅.
> 
> I am willing to try to capture whatever logs I can to help troubleshoot. I haven't used wireshark for a long time and not at all on my current OS. I'll have to figure out how to get it set up and capturing things.

Thanks. But don't worry about it yet. Let me see if I can replicate this first 😀

Cheers,
Martin

**Comment by BiffoBear** [comment]
> > My environment definitely seems to be good at breaking things 😅.
> > I am willing to try to capture whatever logs I can to help troubleshoot. I haven't used wireshark for a long time and not at all on my current OS. I'll have to figure out how to get it set up and capturing things.
> 
> Thanks. But don't worry about it yet. Let me see if I can replicate this first 😀
> 
> Cheers, Martin

@FoamyGuy Hi, I've pushed an update that fixes the TypeError that you saw. Good catch, thank you!

For some reason, my Wiznet5100s cannot close the socket at the end of running `simple_test.py`. The `WIZNET5K` driver sits in an infinite loop waiting for the hardware to report that it has closed the socket. I would like to know if you see the same behaviour?

I've also been spamming `ctrl-C, ctrl-D` on my network without any DHCP errors. When you reset your board, are you also cycling the power to the Wiznet hardware? `ctrl-C, ctrl-D` causes a soft reset of the Wiznet5k, compared to a hard reset from cycling the power. I wonder if this is a factor?

The packets being sent in both cases are correctly formed. The only difference between them is the transaction ID.

If you could run Wireshark for a successful negotiation and an unsuccessful negotiation, it would be a great help, because I will be able to see the UDP headers. As long as the computer running WS is on the same subnet as the Wiznet device it should see the start of the negotiation as this broadcast. Running WS with a `portrange 67-68` will limit the recording to DHCP traffic only.

There's absolutely no urgency about this. 

Cheers,
Martin

P.S.  The exceptions you saw a couple of months ago weren't the off-by-one bug. They were DHCP packets from other transactions being read by the Wiznet (the exception was raised because the transaction ID mismatched). Since these exceptions were being raised consistently, I suspect that there is a lot of DHCP traffic on your network ¯\_(ツ)_/¯.

**Comment by BiffoBear** [comment]
I've somehow got the old version `adafruit_wiznet5k_socket.py` merged into this. Closing the PR while I sort this out.

**Comment by BiffoBear** [comment]
Solved the old `wiznetk5_socket.py` problem. Tested code with `wiznet5k_simpletest.py` and `wiznet5k_simpleserver.py` on a W5100s.

**Comment by FoamyGuy** [comment]
I merged main into this and attempted to resolve the conflicts.

**Comment by BiffoBear** [comment]
I think I've successfully merged the changes. Squished a bug and increased the buffer size from 318 to 512 after receiving a record breaking DHCP response.
DHCP connection successful.
Runs these examples:

1. wiznet5k_simpletest.py
2. wiznet5k_simpleserver.py

**Comment by BiffoBear** [comment]
Great catch on the timeout. That's a bug where I was calculating a timeout based on the value of `time.monotonic`, then later on forgetting that I'd done that and adding `time.monotonic` on again. So that is fixed.

I'm perplexed by your Wireshark logs. Your `wiznet_log_pr_ipsuccess_then_hang.txt` log only contains two discover packets sent by your WIZNET device.
`0000   ff ff ff ff ff ff de ad be ef fe ed 08 00 45 00` bytes 0000 - 0005 are the broadcast MAC address and 0006 - 000c are the sender's MAC address.
Based on this, there shouldn't be an IP address assigned as the DHCP server is never responding.

I've pushed a fix for the timeout bug. I'll spend some time decoding the DHCP packets and get back to you later about that.

Thanks for your review!

**Comment by BiffoBear** [comment]
@FoamyGuy The output from your file `wiznet_log_pr_hangforever_trying_dhcp.txt` contains:
```
Chip Version: w5500
MAC Address: ['0xde', '0xad', '0xbe', '0xef', '0xfe', '0xed']
My IP address is: 192.168.1.208
```
These print statements execute after `eth = WIZNET5K(spi_bus, cs)` and before `print("IP lookup adafruit.com: %s" % eth.pretty_ip(eth.get_host_by_name("adafruit.com")))` so the issue seems to be in the DNS lookup.

I'll dig out a w5500 that I have knocking around and see whether I get the same issue with that. I've been testing on a w5100s.

**Comment by BiffoBear** [comment]
@FoamyGuy I think I've found where it is hanging. I've added a timeout that raises a RuntimeError to break the hang some debug statements that should reveal where the error is occurring. This may not address the underlying cause of the hang though, so I've made the PR a draft for now.

When you have a moment I would appreciate a copy of the serial out with debugging on for a failed connection, in case your wonderfully weird network is doing something different to mine. Thx!

**Comment by FoamyGuy** [comment]
@BiffoBear here are logs from the two different states I'm seeing with the latest version.

This log is after a cold boot up (i.e. unplug / replug) it is partially successful it seems like it got the DNS lookup of adafruit.com done successfully (though outputs as hex instead of dotted string). But then it failed after that with an exception that is raised.
[partial_success_i_think.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11127213/partial_success_i_think.txt)

After that cold boot and partial success if I then do ctrl-C / ctrl-D to restart the script and have it run again I seem to always end up with a "retried 3 times and failed" type error. Here is the log from this cast: 
[failed_after_3_tries.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11127223/failed_after_3_tries.txt)


Let me know if there is anything else that would be helpful to provide or test out on my end. Thank you so much for sticking with this and all of the improvements you've submitted for this library in this PR and beyond!

**Comment by BiffoBear** [comment]
> @BiffoBear here are logs from the two different states I'm seeing with the latest version.
> 
> This log is after a cold boot up (i.e. unplug / replug) it is partially successful it seems like it got the DNS lookup of adafruit.com done successfully (though outputs as hex instead of dotted string). But then it failed after that with an exception that is raised. [partial_success_i_think.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11127213/partial_success_i_think.txt)
> 
> After that cold boot and partial success if I then do ctrl-C / ctrl-D to restart the script and have it run again I seem to always end up with a "retried 3 times and failed" type error. Here is the log from this cast: [failed_after_3_tries.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11127223/failed_after_3_tries.txt)
> 
> Let me know if there is anything else that would be helpful to provide or test out on my end. Thank you so much for sticking with this and all of the improvements you've submitted for this library in this PR and beyond!

Thanks, that partial success showed me where my logic was flawed. I've pushed a band-aid fix. Once this PR is done, I've got some ideas for the `WIZNET5K` class.

I don't see why there is apparently no reply from the DHCP server after a warm reboot. That really puzzles me.

Please would you record the exchange in Wireshark? If you use the capture filter `portrange 67-68` that will only capture DHCP packets. Once the DHCP fails after 3 retries please stop the capture and save the results as a `Wireshark/… - pcapng` file. I can open this file in Wireshark here and get a better idea of the exchange with the server. Maybe my PR is missing valid packets from the server?

Thanks for all your help and encouragement. No rush.

Martin

**Comment by BiffoBear** [comment]
@FoamyGuy  I may have been barking up the wrong tree. I've added an explicit reset to the w5500 detection code. When you have a moment, please see if this solves your no DHCP after soft reset issue. Thanks!

**Comment by FoamyGuy** [comment]
@BiffoBear I have tested out the latest versions

Unfortunately the latest commit seems to cause more issues in my environment. If I test the absolute latest version including the newest commit I consistently end up with `Failed to initialize WIZnet module.` exception raised. I tried many times both from cold boot and after reset and always had the same results.  My instinct was that perhaps we needed to sleep some time after the reset command before attempting anything else. I did tinker with adding a sleep, but it didn't seem to have any effect. I still kept getting failure to initialize.

In my local copy I commented out the lines added in the latest commit so that I could test the version from before that and capture some additional logs.

All of this output was obtained on a Feather S2 TFT with the latest version minus the last commit with the reset command, that is commented out in the version running while these logs were captured:

**debug=True logs from serial console:**
[debug_dhcp_failed_from_reset_after_success.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301746/debug_dhcp_failed_from_reset_after_success.txt)
[debug_full_success_from_cold_boot.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301747/debug_full_success_from_cold_boot.txt)

**wireshark logs from failed DHCP attempt:**
[wireshark_fail_dhcp_0.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301748/wireshark_fail_dhcp_0.txt)
[wireshark_fail_dhcp_1.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301749/wireshark_fail_dhcp_1.txt)
[wireshark_fail_dhcp_2.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301750/wireshark_fail_dhcp_2.txt)
[wireshark_fail_dhcp_3.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301751/wireshark_fail_dhcp_3.txt)

**Wireshark logs from a successful DHCP and simpletest exectution:**
[wireshark_success_from_cold_boot_0.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301752/wireshark_success_from_cold_boot_0.txt)
[wireshark_success_from_cold_boot_1.txt](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/files/11301753/wireshark_success_from_cold_boot_1.txt)


My results with this version seem to be consistently that after a cold boot up it will successfully connect, get an IP, and successfully fetch data from the internet.  Then if I ctrl-D / ctrl-C to reset it will consistently fail with `No response from DHCP server after 3 retries.` Resetting subsequent times doesn't change the behavior seemingly it's always this 'no response' until power cycled.

After a full power cycle it boots up and succeeds fully 1 time again, then upon first reset is back to the 'no response' state.

**Comment by BiffoBear** [comment]
OK. I'll save it as a draft and wait until I can get my hands on a W5500 to test against Thanks for checking!

**Comment by BiffoBear** [comment]
@FoamyGuy This time I promise 🤪! I fixed my w5500 so I could test on it. On the current main branch I get the error after 5 x ctrl-D with a w5500, never with a w5100s. On the latest commit for this PR I don't get any errors after 10 x ctrl-D on the W5500 or the w5100s.

**Comment by FoamyGuy** [comment]
@BiffoBear nice! 

I gave the latest version a try and I am getting continued success with it after several resets, running the simpletest. 

I'm thinking this is likely good now, but I am going to plan to come back once more and do some more thorough testing with more of the examples. 

Thank you so much for seeing it through and making all of these improvements!

**Comment by BiffoBear** [comment]
> @BiffoBear nice!
> 
> I gave the latest version a try and I am getting continued success with it after several resets, running the simpletest.
> 
> I'm thinking this is likely good now, but I am going to plan to come back once more and do some more thorough testing with more of the examples.
> 
> Thank you so much for seeing it through and making all of these improvements!

@FoamyGuy And thank you for being such a patient, insightful and thorough reviewer.

**Review by FoamyGuy — APPROVED** [review]
These changes are looking good to me at this point. 

I tested successfully several of the examples on a Feather RP2040 v8.1.0-beta.2. Everything is functioning as intended and seems to run reliably even through several soft resets.

Thanks again for these improvements @BiffoBear!

**Pull request merged**